gpflow/models/training_mixins.py

```diff
@@ -0,0 +1,58 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar, Union
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData, OutputData)
+
+
+class InternalDataTrainingLossMixin:
+    def training_loss(self) -> tf.Tensor:
+        """
+        Training loss for models that encapsulate their data.
+        """
+        return -self.maximum_a_posteriori_objective()
+
+    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
+        if jit:
+            return tf.function(self.training_loss)
+        return self.training_loss
+
+
+class ExternalDataTrainingLossMixin:
+    def training_loss(self, data: RegressionData) -> tf.Tensor:
+        """
+        Training loss for models that do not encapsulate the data.
+        
+        :param data: the data to be used for computing the model objective.
+        """
+        return -self.maximum_a_posteriori_objective(data)
+
+    def training_loss_closure(
+        self, data: Union[Data, collections.abc.Iterator], jit=True
```

Comment:
will this work for all `Iterators`?

Reply:
as an aside, this should perhaps be `typing.Iterator`

Reply:
We could change the type signature, I just thought it might as well be consistent with the `isinstance` check in the function body.

Reply:
yeah let's do that

Reply:
done